### ISSUE TITLE
[PATCH] bmodpow() fails when GMP library is used.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11093]
Migrated from [rt.perl.org#82966](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=82966) (status was 'resolved')

Searchable as RT82966$

**Comment by p5pRT** [comment]
### From @pjacklam

This bug is not seen when the libraries Calc and FastCalc are used\,
because their \_modpow\(\) method modifies the first argument\. However\,
the GMP library's \_modpow\(\) does not modify the first argument\,
causing a so far undetected bug to show up and fail tests\.

Using the Calc library prints the correct "\-4"&#8203;:

&nbsp;    use Math&#8203;::BigInt lib => Calc;
&nbsp;    $x = Math&#8203;::BigInt\->new\(8\);
&nbsp;    $y = Math&#8203;::BigInt\->new\(8\);
&nbsp;    $z = Math&#8203;::BigInt\->new\(\-5\);
&nbsp;    print $x \-> bmodpow\($y\, $z\)\, "\\n";

Using the GMP library prints the incorrect "\-\-"&#8203;:

&nbsp;    use Math&#8203;::BigInt lib => GMP;
&nbsp;    $x = Math&#8203;::BigInt\->new\(8\);
&nbsp;    $y = Math&#8203;::BigInt\->new\(8\);
&nbsp;    $z = Math&#8203;::BigInt\->new\(\-5\);
&nbsp;    print $x \-> bmodpow\($y\, $z\)\, "\\n";


---
 dist/Math\-BigInt/lib/Math/BigInt\.pm |   15 \+\+\+\+\+\+\+\+\-\-\-\-\-\-\-
 1 files changed\, 8 insertions\(\+\)\, 7 deletions\(\-\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/Math-BigInt/lib/Math/BigInt.pm b/dist/Math-BigInt/lib/Math/BigInt.pm
index 45a0f01..177894a 100644
--- a/dist/Math-BigInt/lib/Math/BigInt.pm
+++ b/dist/Math-BigInt/lib/Math/BigInt.pm
@@ -18,7 +18,7 @@ package Math::BigInt;
 my $class = "Math::BigInt";
 use 5.006002;
 
-$VERSION = '1.99_04';
+$VERSION = '1.99_05';
 
 @ISA = qw(Exporter);
 @EXPORT_OK = qw(objectify bgcd blcm); 
@@ -1858,9 +1858,9 @@ sub bmodinv
 
 sub bmodpow
   {
-  # takes a very large number to a very large exponent in a given very
-  # large modulus, quickly, thanks to binary exponentiation. Supports
-  # negative exponents.
+  # Modular exponentiation. Raises a very large number to a very large exponent
+  # in a given very large modulus quickly, thanks to binary exponentiation.
+  # Supports negative exponents.
   my ($self,$num,$exp,$mod,@r) = objectify(3,@_);
 
   return $num if $num->modify('bmodpow');
@@ -1890,7 +1890,7 @@ sub bmodpow
   # If the resulting value is non-zero, we have four special cases, depending
   # on the signs on 'a' and 'm'.
 
-  unless ($CALC->_is_zero($num->{value})) {
+  unless ($CALC->_is_zero($value)) {
 
       # There is a negative sign on 'a' (= $num**$exp) only if the number we
       # are exponentiating ($num) is negative and the exponent ($exp) is odd.
@@ -1913,7 +1913,7 @@ sub bmodpow
           else {
               # Use copy of $mod since _sub() modifies the first argument.
               my $mod = $CALC->_copy($mod->{value});
-              $value = $CALC->_sub($mod, $num->{value});
+              $value = $CALC->_sub($mod, $value);
               $sign  = '+';
           }
 
@@ -1925,8 +1925,9 @@ sub bmodpow
           #               = -(m - (a (mod m)))
 
           if ($mod->{sign} eq '-') {
+              # Use copy of $mod since _sub() modifies the first argument.
               my $mod = $CALC->_copy($mod->{value});
-              $value = $CALC->_sub($mod, $num->{value});
+              $value = $CALC->_sub($mod, $value);
               $sign  = '-';
           }
 
-- 
1.7.3.3
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Jan 28 09&#8203;:48&#8203;:20 2011\, pjacklam wrote&#8203;:
> This bug is not seen when the libraries Calc and FastCalc are used\,
> because their \_modpow\(\) method modifies the first argument\. However\,
> the GMP library's \_modpow\(\) does not modify the first argument\,
> causing a so far undetected bug to show up and fail tests\.

Thank you\. I have just applied this as 6c29054\. Does this fix \#81770?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @pjacklam

On Fre\. 28\. Jan\. 2011 20&#8203;:21&#8203;:33\, sprout wrote&#8203;:

> Does this fix \#81770?

Father Chrysostomos&#8203;:

> Thank you\. I have just applied this as 6c29054\. Does this fix \#81770?

There is no need to do more work on Math\-BigInt to fix this issue\, but
there is still a problem in the Math\-BigInt\-GMP distribution\. I am
working on it\, and will release a new version of Math\-BigInt\-GMP within
a few days\.

The background for many of the recent Math&#8203;::BigInt problems is that I
fixed Math&#8203;::BigInt using the Math&#8203;::BigInt&#8203;::Calc library\, assuming that
the various libraries \(Math&#8203;::BigInt&#8203;::Calc\, Math&#8203;::BigInt&#8203;::FastCalc\,
Math&#8203;::BigInt&#8203;::GMP\, \.\.\.\) behave the same way\. It turns out that they
don't\. Not surprisinly\, as there is no explicit testing of the API\.
Actually\, there is not even an unambiguous description of the API\. I am
now patching Math&#8203;::BigInt&#8203;::GMP so it works with the most recent
Math&#8203;::BigInt\.
